---
title: "MISTy applied to Spatial Transcriptomics"
author: "Ricardo Ramirez & Jovan Tanevski"
date: "7/30/2020"
output: html_document
---

In this notebook we show how MISTy can be used to study the influence of tissue organization in the expression of markers.

#### 1. Loading packages

```{r setup, include=T,warning=F,message=F}
library(OmnipathR)
library(tidyverse)
library(Seurat)
library(cowplot)
library(MISTy)
library(future)
```

#### 2. Loading spatial data

We exemplify usage using Seurat objects, however, MISTy is not dependent on this.

In our case, pathway activities are stored in the *progeny* assay and gene expression in the *SCT* assay

```{r,eval=F}
visium_slide = readRDS(file = "./A1_out/breast_A_1.rds")

print(Assays(visium_slide))
```

#### 3. Defining MISTy pipelines

MISTy models are built around a prediction task. For each marker of interest in a sample, we can model spot-spot interactions coming from different spatial contexts as different views. The first and main view, containing all markers of interest, is the intrinsic view, where we relate the expression of other markers to a specific marker of interest within the same location. To capture the local cellular niche, we can create a view that relates the expression from the immediate neighborhood of a cell to the observed expression within that cell; we call this view a juxtaview. To capture the effect of the tissue structure, we can create a view that relates the expression of markers measured in cells within a radius around a given cell, and we call this view a paraview

Each MISTy pipeline requires 3 inputs:\ 
1) features to predict (can be genes, pathway activities, etc),\   
2) predictors and\
3) spots positions to study\  

In the following pipeline. We build a default model that uses an intrinsic view and a paracrine view. Note that MISTy is not limited by this and multiple views can be added as neccesary using `add_paraview`, `add_juxtaview`, `create_view`, and `add_views`.

```{r,eval=F}
#' Runs MISTy classic paraview pipeline 
#' 
#' Warning: Feature IDs can't have "-" symbol
#' Warning: rownames must be the same in intra_df, para_df and geometry
#'
#' @param intra_df: feature data_frame with IDs as rows and features as columns
#' @param intra_features: features of the intraview to be used, if NULL uses all columns. These are the predicted markers
#' @param para_df: feature data_frame with IDs as rows and features as columns
#' @param para_features: features of the paraview to be used, if NULL uses all columns. These are the predictors
#' @param geometry: a data frame with IDs as rows two coordinates as columns 
#' @param l: radius parameter
#' @param spot_ids: spot IDs to fit MISTy if null all rows are used. This only affects the model fit and not the view calculation.
#' @param out_alias: folder name to be used in all MISTy outputs
para_pipeline = function(intra_df, 
                         intra_features = NULL,
                         para_df,
                         para_features = NULL,
                         geometry,
                         l,
                         spot_ids = NULL,
                         out_alias = "default"){
  
  plan(multiprocess, workers = 4)
  
  clear_cache()
  
  if(is.null(intra_features)){
    intra_features = colnames(intra_df)
  }
  
  if(is.null(para_features)){
    para_features = colnames(para_df)
  }
  
  if(is.null(spot_ids)){
    spot_ids = rownames(intra_df)
  }
  
  # Defining useful data intra
  intra_df = intra_df[spot_ids,intra_features]
  colnames(intra_df) = gsub("-","_", colnames(intra_df))
  
  views_main = create_initial_view(intra_df, 
                                   unique.id = "intra")
  
  
  # Defining useful data para
  para_df = para_df[rownames(geometry),para_features]
  
  colnames(para_df) = gsub("-","_", colnames(para_df))
  
  views_para = create_initial_view(para_df, 
                                   unique.id = paste0("para_",l^2)) %>% 
    add_paraview(geometry, l^2)
  
  # Fetching actual para info to be used
  
  # Spot specific view comes from the view above
  data_red = views_para[[3]]$data
  rownames(data_red) = rownames(para_df) #we named rows just for easy access
  data_red = data_red[rownames(intra_df),]
  
  # Define frankenstein of views
  views = views_main %>% 
    add_views(create_view(paste0("para_",l^2),
                          data_red))
  
  MISTy_run = run_misty(views,paste0(out_alias,"_",l^2))
  
}

#' Runs MISTy classic paraview pipeline in Seurat objects
#'
#' @param visium_slide: visium seurat object
#' @param intra_assay: assay where to extract the predicted markers
#' @param intra_features: features of the intraview to be used, if NULL uses all columns
#' @param para_assay: assay where to extract the predictor markers
#' @param para_features: features of the paraview to be used, if NULL uses all columns
#' @param l: radius parameter
#' @param spot_ids: spot IDs to fit MISTy if null all rows are used
#' @param out_alias: folder name to be used in all MISTy outputs

para_ppln_seurat = function(visium_slide,
                                      intra_assay, 
                                      intra_features = NULL,
                                      para_assay,
                                      para_features = NULL,
                                      l,
                                      spot_ids = NULL,
                                      out_alias = "default"){
  
  # Getting data ready to create views
  geometry = visium_slide@images$slice1@coordinates[,c(2,3)]
  
  # Intrinsic data
  intra_df = as.matrix(visium_slide@assays[[intra_assay]]@data)
  intra_df = intra_df %>%
    t %>% data.frame(check.names = F)
  
  intra_df = intra_df[rownames(geometry),]
  
  # Para data
  para_df = as.matrix(visium_slide@assays[[para_assay]]@data)
  para_df = para_df %>%
    t %>% data.frame(check.names = F)
  para_df = para_df[rownames(geometry),]
  
  
  para_pipeline(intra_df = intra_df,
                intra_features = intra_features,
                para_df = para_df,
                para_features = para_features,
                geometry = geometry,
                l = l,
                spot_ids = spot_ids,
                out_alias = out_alias)
  
}
```

#### 4. Running the pipeline in the visium dataset

A first interest is to predict pathway activities with putative ligands extracted from Omnipath

```{r,eval=F}
lig_rec = import_intercell_network(
  interactions_param = list(datasets = c('ligrecextra', 'omnipath', 'pathwayextra'),
                            transmitter_param = list(parent = 'ligand'),
                            receiver_param = list(parent = 'receptor')
  ))

ligands = unique(lig_rec$source_genesymbol)
```

We filtered out all ligands that have a coverage of less than 30% in all the slide

```{r,eval=F}
#Useful ligands
slide_ligands = ligands[ligands %in% rownames(visium_slide)] 
gex = as.matrix(visium_slide@assays$SCT@data)[slide_ligands,]
lig_coverage = rowSums(gex>0)/ncol(gex)
slide_ligands = names(lig_coverage[lig_coverage>=.3])
```

Running MISTy

```{r,eval=F}
ls = c(5)
misty_out_path = "./A1_out/misty_progeny_ligands" # Defining the output folder alias

test_opath = lapply(ls,para_ppln_seurat,
                    visium_slide = visium_slide,
                    intra_assay = "progeny",
                    intra_features = NULL,
                    para_assay = "SCT",
                    para_features = slide_ligands,
                    spot_ids = NULL,
                    out_alias = misty_out_path)
```

#### 5. Aggregation

MISTy's output folder contains the following files:\n

**performance** : Contains the $R^2$ and $RMSE$ of the intraview and multiview model. A p-value of the difference in these measurements between the two models is also provided, and reflects if the additional views improved the performance. This is done by performing a t-test on vectors of samples of $R^2$ and $RMSE$ coming from crossvalidation. The p-value is used to weight the predictor importances too.\
**coefficients** : The results of the linear model that combines the predictions of each view and are used to measure the contribution of each view.\
**importances** : Predictor importances for each target marker.\

Below we provide a function to integrate all these files in a final R object:

```{r}

#'Function to get processed MISTys final output
#'@param results_folder: result folder from a MISTy run or an optimization search
#'@param p.cutoff: p-value cutoff to evaluate coefficients
#'
#'@return a list with improvement, contribution, coefficients and importance info
MISTy_aggregator = function(results_folder,
                            p.cutoff = 0.05){
  
  images = results_folder
  
  # Improvement 
  impr = images %>% map_dfc(function(image) {
    performance <- read_delim(paste0(image, .Platform$file.sep,
                                     "performance.txt"),
                              delim = " ", col_types = cols()
    ) %>% distinct()
    
    targets <<- unique(performance$target)
    performance %>%
      arrange(target) %>%
      transmute(RMSE = (intra.RMSE - multi.RMSE) / intra.RMSE, R2
                = (multi.R2 - intra.R2))
  })
  
  avg <- ((images %>% map(function(image) {
    coefficients <- read_delim(paste0(image, .Platform$file.sep, "coefficients.txt"),
                               delim = " ", col_types = cols()
    )
    
    targets <<- coefficients %>%
      pull(target) %>%
      sort
    
    coefficients %>%
      distinct() %>%
      arrange(target) %>%
      dplyr::select(-target, -contains("intercept")) %>%
      mutate_at(vars(starts_with("p.")), ~ as.numeric(. <= p.cutoff)) %>%
      mutate_at(vars(-starts_with("p.")), abs)
  }) %>% 
    purrr::reduce(`+`)) / length(images)) %>% 
    mutate(target = targets)
  
  ctotals <- avg %>% 
    dplyr::select(-starts_with("p."), -"target") %>%
    rowSums
  
  coefs <- avg %>% 
    dplyr::select(-starts_with("p.")) %>%
    mutate_if(is.numeric, ~./ctotals) %>%
    tidyr::pivot_longer(-target, names_to = "view")
  
  maps = images %>% map(function(image) {
    coefficients <- read_delim(paste0(image, .Platform$file.sep, "coefficients.txt"),
                               delim = " ", col_types = cols()
    ) %>% distinct()
    
    targets <- unique(coefficients$target)
    views <<- (coefficients %>% dplyr::select(-target, -starts_with("p."), -intercept) %>% colnames())
    
    # one heatmap per view
    maps <- views %>% map(function(view) {
      all.importances <- targets %>% map(~ read_csv(paste0(
        image, .Platform$file.sep, "importances_",
        .x, "_", view, ".txt"
      ),
      col_types = cols()
      ) %>%
        distinct() %>%
        filter(!grepl("_2$", target)))
      
      features <- unique(all.importances %>% map(~ .x$target) %>% unlist())
      
      pview <- paste0("p.", view)
      ps <- coefficients %>%
        dplyr::select(target, !!pview) %>%
        mutate(!!pview := (1 - !!sym(pview)))
      
      
      # importances are standardized for each target an multiplied by 1-pval(view)
      result <- all.importances %>%
        imap_dfc(~
                   tibble(target = features, zero.imp = 0) %>%
                   left_join(.x, by = "target") %>%
                   transmute(feature = target, importance = (zero.imp + scale(imp)[, 1]) *
                               (ps %>% filter(target == targets[.y]) %>% pull(pview))) %>%
                   dplyr::select(importance)) %>%
        `colnames<-`(targets) %>%
        mutate(Predictor = features)
      
      # in order for aggregation
      result %>%
        arrange(Predictor) %>%
        dplyr::select((order(colnames(result))))
      #dplyr::select(noquote(order(colnames(result))))
    })
  })
  
  aggregated = maps %>% purrr::reduce(function(acc, l) {
    map2(acc, l, ~ (((.x %>% dplyr::select(-Predictor)) + (.y %>% dplyr::select(-Predictor))) %>%
                      mutate(Predictor = .x %>% pull(Predictor))))
  })
  
  performance_df = read_delim(paste0(results_folder,"/performance.txt"), delim = " ")
  
  result_list = list("impr" = impr,
                     "targets" = targets,
                     "coefs" = coefs,
                     "importance" = aggregated,
                     "performance" = performance_df)
  
  return(result_list)
  
}
```

Running the functions in our output file:

**Beware** that given the parallelization, the coefficients or performance files may have skipped return lines.

```{r}
folders = "./A1_out/misty_progeny_ligands_25"
MISTy_out = MISTy_aggregator(results_folder = folders)
```

#### 6. Visualization

Here we provide some functions to summarize the results.

**Plotting performance**

```{r}
#'Plot MISTy performance
#'@param MISTy_out: result list from MISTy_aggregator
#'@param pdf_out: pdf file where to return plots
#'@param predicted_features: show results of selected features, if NULL all
#'
#'@return a pdf with performance plots

plot_misty_performance = function(MISTy_out,
                                  predicted_features = NULL){
  #Generating figures
  #Overall performance
  performance_df = MISTy_out$performance %>% tidyr::pivot_longer(cols = -target) %>%
    dplyr::filter(grepl("R2",name))
  
  R2_impr = MISTy_out$impr %>% dplyr::select(contains("R2")) %>%
    mutate(target = targets) %>%
    pivot_longer(cols = -target, 
                 names_to = "name", 
                 values_to = "value") %>% 
    arrange(desc(value))
  
  #Look at importances of views
  contribution_df = MISTy_out$coefs
  
  if(length(predicted_features)>0){
    
    performance_df = performance_df %>%
      dplyr::filter(target %in% predicted_features)
    
    R2_impr = R2_impr %>%
      dplyr::filter(target %in% predicted_features)
    
    contribution_df = contribution_df %>%
      dplyr::filter(target %in% predicted_features)
    
  }
  
  performance_plt = ggplot(performance_df,
                             aes(fill = name, y = target, x = value)) +
    geom_bar(stat = "identity",position="dodge") + 
    theme_minimal()
  
  impr_plot = ggplot(R2_impr) +
    geom_point(aes(x = target, y = value * 100)) +
    theme_classic() +
    ylab("Change in variance explained") +
    xlab("Target") +
    theme(axis.title = element_text(size=11),
          axis.text = element_text(size=10),
          axis.text.x = element_text(angle = 90, hjust = 1))
  
  coefs_plot = ggplot(contribution_df) + 
    geom_col(aes(x=target, 
                 y=value, group=view, fill=view)) +
    xlab("Target") +
    ylab("Contribution") +
    theme_classic() +
    theme(axis.text = element_text(size=13),
          axis.title = element_text(size=13),
          legend.text = element_text(size=11)) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1,
                                     vjust = 0.5)) +
    scale_fill_brewer(palette="Dark2",
                      labels = c("Intrinsic","Para pathway")) +
    labs(fill = "View")
  
  plot(performance_plt)
  plot(impr_plot)
  plot(coefs_plot)
}
```

```{r}
plot_misty_performance(MISTy_out = MISTy_out)
```

**Plotting importances**

In these functions, we only keep predictors with an importance value over a cutoff in at least one target marker.

```{r}
#' Clusterize importance hmaps from MISTy
#' @param importance_obj : Specific importance tibble to cluster
#' @param predictors_vars : predictor variables to consider in the clustering
#' @param predicted_vars : predicted variables to consider in the clustering
#' @return a list with the order of predictors and predicted variables
cluster_importance = function(importance_obj,
                              predictors_vars,
                              predicted_vars){
  
  importance_intra_mat = importance_obj %>% 
    dplyr::filter(Predictor %in% predictors_vars) %>%
    as.data.frame()
  
  rownames(importance_intra_mat) = importance_intra_mat$Predictor
  importance_intra_mat = as.matrix(importance_intra_mat[,-which(colnames(importance_intra_mat) == "Predictor")])
  importance_intra_mat = importance_intra_mat[,predicted_vars]
  
  predictor_clust =  hclust(d = t(dist(importance_intra_mat)))
  predicted_clust =  hclust(d = t(dist(t(importance_intra_mat))))
  
  return(list("Predictor_order" = predictor_clust$labels[predictor_clust$order],
         "Predicted_order" =  predicted_clust$labels[predicted_clust$order]))
  
  
}

#'Plot view importances: To do, expand to multiple views with a walk through views
#'@param MISTy_out: result list from MISTy_aggregator
#'@param pdf_out: pdf file where to return plots
#'@param importance_cut: to filter predictors of no use
#'@return a pdf document with importance plots
#'
#'TO DO, make it independent so that it is applied to the list in MISTy_out
plot_misty_importance = function(MISTy_out,
                                 predicted_features = NULL,
                                 predictors_features = NULL,
                                 importance_cut = 1,
                                 make_clust = T,
                                 width_pdf = 15,
                                 height_pdf = 13){
  
  # We have a list of importance heatmaps in the MISTy_out file
  n_views = length(MISTy_out$importance)
  importance_out = list()
  
  for(i in 1:n_views){
    
    importance_df = tidyr::gather(MISTy_out$importance[[i]], 
                                  "Predicted",
                                  "Importance", -Predictor)
    
    predicted_features_v = predicted_features
    
    predictors_features_v = predictors_features[[i]]
    
    if(length(predicted_features_v)>0){
      importance_df = importance_df %>% 
        dplyr::filter(Predicted %in% predicted_features_v)
    }
    
    if(length(predictors_features_v)>0){
      importance_df = importance_df %>% 
        dplyr::filter(Predictor %in% predictors_features_v)
    }
    
    #Here we extract predictors that have a predictor importance of at least importance_cut
    
    importance_summ = importance_df %>% 
      dplyr::mutate(importance_bool = Importance >= importance_cut) %>%
      group_by(Predictor) %>% 
      summarize(predictor_summ = sum(importance_bool,na.rm = T)) %>%
      dplyr::filter(predictor_summ >= 1)
    
    importance_predictors = importance_summ %>% select(Predictor) %>% pull()
    
    importance_df = importance_df %>% 
      dplyr::filter(Predictor %in% importance_predictors) %>%
      dplyr::arrange(Predicted, -Importance)
    
    # Clustering the plots
    
    if(make_clust){
      clust_order = cluster_importance(importance_obj = MISTy_out$importance[[i]],
                                       predictors_vars = unique(importance_predictors),
                                       predicted_vars = unique(importance_df$Predicted))
      
      imp_plot = importance_df %>% 
        ggplot(aes(x = factor(Predictor,
                              levels = clust_order$Predictor_order), 
                   y = factor(Predicted,
                              levels = clust_order$Predicted_order) ,
                   fill = Importance)) + geom_tile() + 
        theme(panel.grid = element_blank(),
              axis.title = element_text(size=11),
              axis.text.x = element_text(angle = 90,
                                         hjust = 1,
                                         vjust = 0.5),
              axis.text = element_text(size=10),
              legend.key.size = unit(.6, "cm"),
              legend.title = element_text(size = 10),
              legend.text = element_text(size = 10),
              legend.position = "bottom") +
        scale_fill_gradient2(low = "white", 
                             mid = "white", 
                             high = scales::muted("blue"),
                             midpoint = 0.9) +
        xlab("Predictor features") + ylab("Predicted markers")
    }else{
      
      imp_plot = importance_df %>% 
        ggplot(aes(x = Predictor, 
                   y = Predicted,
                   fill = Importance)) + geom_tile() + 
        theme(panel.grid = element_blank(),
              axis.title = element_text(size=11),
              axis.text.x = element_text(angle = 90,
                                         hjust = 1,
                                         vjust = 0.5),
              axis.text = element_text(size=10),
              legend.key.size = unit(.6, "cm"),
              legend.title = element_text(size = 10),
              legend.text = element_text(size = 10),
              legend.position = "bottom") +
        scale_fill_gradient2(low = "white", 
                             mid = "white", 
                             high = scales::muted("blue"),
                             midpoint = 0.9) +
        xlab("Predictor features") + ylab("Predicted markers")
      
    }
    
    plot(imp_plot)
    
    importance_out[[i]] = importance_df
    
  }
  return(importance_out)
}
```


```{r}
plot_misty_importance(MISTy_out,
                      importance_cut = 1)
```
